compute/cker/src/train/DepthwiseConv.test.cc

```diff
@@ -0,0 +1,410 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cker/train/operation/DepthwiseConv.h>
+
+#include <gtest/gtest.h>
+#include <vector>
+
+namespace
+{
+
+template <typename T> class DepthwiseConvVerifier
+{
+public:
+  DepthwiseConvVerifier() : _dconv_kernel{new nnfw::cker::train::DepthwiseConv()}
+  {
+    _dconv_kernel = std::make_unique<nnfw::cker::train::DepthwiseConv>();
+  }
+
+  void prepare(const nnfw::cker::Shape &incoming_shape, const nnfw::cker::Shape &filter_shape)
+  {
+    const int kPacketSize = _dconv_kernel->kPacketSize<T>();
+    const int batch = incoming_shape.Dims(0);
+    const int out_depth = incoming_shape.Dims(3);
+    const int filter_rows = filter_shape.Dims(1);
+    const int filter_cols = filter_shape.Dims(2);
+    const int filter_spatial_size = filter_rows * filter_cols;
+    const int padded_filter_inner_dim_size =
+      ((out_depth + kPacketSize - 1) / kPacketSize) * kPacketSize;
+
+    _use_padded_filter = (out_depth % kPacketSize) == 0 ? false : true;
+    {
+      nnfw::cker::Shape padded_filter_shape(
+        {batch, filter_spatial_size, padded_filter_inner_dim_size});
+      _padded_filter.resize(padded_filter_shape.FlatSize());
+    }
+
+    {
+      const int thread_count = _dconv_kernel->getThreadCount() + 1;
+
+      nnfw::cker::Shape filter_buffer_shape(
+        {thread_count, filter_spatial_size, padded_filter_inner_dim_size});
+      _filter_buffers.resize(filter_buffer_shape.FlatSize());
+
+      nnfw::cker::Shape filter_dim_buffer_shape({thread_count, padded_filter_inner_dim_size});
+      _filter_dim_buffers.resize(filter_dim_buffer_shape.FlatSize());
+    }
+  }
+
+  void verifyInputGradExpected(const nnfw::cker::DepthwiseConvParams &params,
+                               const nnfw::cker::Shape &incoming_shape, const T *incoming_data,
+                               const nnfw::cker::Shape &filter_shape, const T *filter_data,
+                               const nnfw::cker::Shape &grad_shape)
+  {
+    std::vector<T> gradient(grad_shape.FlatSize(), static_cast<T>(0));
+    std::vector<T> expected(grad_shape.FlatSize(), static_cast<T>(0));
+
+    calculateInputGradExpected(params, incoming_shape, incoming_data, filter_shape, filter_data,
+                               grad_shape, expected.data());
+
+    _dconv_kernel->backpropInput(params, incoming_shape, incoming_data, filter_shape, filter_data,
+                                 _padded_filter.data(), grad_shape, gradient.data(),
+                                 _use_padded_filter, _filter_buffers.data(),
+                                 _filter_dim_buffers.data());
+
+    for (size_t i = 0; i < gradient.size(); ++i)
+      EXPECT_NEAR(gradient[i], expected[i], 1e-3f);
```

Comment:
There is no checking `_filter_buffers` and `_filter_dim_buffers` even if they are updated. Could you add the verification to check them?

Reply:
This buffers are used for temporary memory. Should I verify the temp memory?

Reply:
It seems that there was a misunderstanding about what we talked about offline in the past. I think It's better to manage temporary memories in kernel if they are only used in kernel?
Could you move temporary memories into kernel?

Reply:
I don't understand what you said. Why do you suggest moving these temporary memories into the kernel? If these memories are located in the backend kernel, we can reuse these temporary memory for optimization.

Reply:
As we talked about offline, I suggested moving those memories into the kernel since those memories are only used in kernel. But I couldn't find the way for cker to manage and optimize temporary memories. I think temporary memories would be managed directly in train backend. In short, it would be better to keep it as is.